Advanced search step2

asaunier opened the pull request:
Related to https://github.com/c2corg/v6_ui/issues/368

TODO

* [x] take "more filters" filters into account for waypoints. As for now only the first-level criterias are taken into account (title, waypoint_types, elevation).
* [ ] filters for routes
* [ ] filters for outings
* ~~cards are not rendered correctly in debug mode (same issue than for the simple search)~~ => see https://github.com/c2corg/v6_ui/issues/369
* [x] outing cards layout need some CSS fixing
* [x] feature icon is now clickable in the detail view page map: either disable this behaviour or at least make sure the URL is OK
* [x] add an extern entry for the ``listDocuments`` function in ``apiService``.
* [x] make sure the features on the map use the most appropriate locales.
* [ ] Skip the map extent when filtering on a WP or a route ID? (with permalink like ``outings?r=123456``)
* [x] Handle orientations
* [x] Use sliders for ratings (ie. non-numeric values) as well?
* [ ] Handle date filters

TOFIX
* [ ] slider handles are not fully reset when clearing filters
* [x] loading spinner no longer shows up when changing the filters!?

EXTRA

* [ ] Hilight the feature on the map when hovering a card (and vice-versa?)
* [ ] allow to bypass the map extent filtering as in Airbnb?

Some changes in the API are needed: https://github.com/c2corg/v6_api/issues/311

gberaudo commented on c2corg_ui/static/js/card.js:
```diff
@@ -61,17 +60,23 @@ app.CardController = function(gettextCatalog) {
-  this.locale = this['doc']['locales'][0];
-  for (var i = 0, n = this['doc']['locales'].length; i < n; i++) {
-    var l = this['doc']['locales'][i];
+  this.locale = this.doc['locales'][0];
```
`locales` is a property of the document which is defined in the extern so you can also remove these brackets.

ginold replied:
so I  too was doing it wrong my whole life ...

asaunier replied:
It seems to work indeed. But I am not sure to understand why because I don't really see the typedef for documents in the externs https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/externs/appx.js

gberaudo replied:
https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/externs/appx.js#L38-L47

gberaudo commented on c2corg_ui/static/js/card.js:
```diff
@@ -95,9 +100,7 @@ app.CardController.prototype.translate = function(str) {
-    app.utils.getDoctype(this.doc_['type']),
-    this.doc_['document_id'],
-    this.doc_['locales'][0]['lang']);
+    this.type, this.doc['document_id'], this.doc['locales'][0]['lang']);
```
Idem, the `document_id` brackets can be removed.

asaunier commented:
This PR is growing really big again :P 
Actually it contains most of the change for the advanced search pages.

The outings and routes pages are not ready yet (not the right filters) but at least they work as in alpha1. I could add the filters in some other PR. There are also a few other remaining points in the TODO/TOFIX list but not critical.

Feel free to review, thanks :)

tsauerwein commented on c2corg_ui/static/js/card.js:
```diff
@@ -67,12 +65,18 @@ app.CardController = function(gettextCatalog) {
-    var l = this.doc['locales'][i];
+  this.locale = this.doc.locales[0];
+  for (var i = 0, n = this.doc.locales.length; i < n; i++) {
+    var l = this.doc.locales[i];
```
When using `.locales`, you should type `this.doc`.

asaunier replied:
You're definitely right. I have just made a change with 55d2eb4

It seemed to work anyway since I previously had no warning when compiling and the cards locales were correctly fed.

tsauerwein commented on c2corg_ui/static/js/search/searchfilters.js:
```diff
@@ -189,4 +223,42 @@ app.SearchFiltersController.prototype.selectOption = function(prop, val, event)
+  } else {
+    this.orientations = this.orientations.filter(function(val) {
+      return val !== orientation;
+    });
```
I don't understand what you are doing with the filter. You are creating a new array that contains all orientations in `this.orientations` that are not `orientation`. But in the `if` above you have already checked that `orientation` is not in `this.orientations`?

asaunier replied:
When using this ``setOrientation`` method (triggered when clicking on one of the directions of the SVG compass, eg. at https://github.com/c2corg/v6_ui/blob/advanced-search-step2/c2corg_ui/templates/helpers/orientations.svg?short_path=31b8504#L62):
* either the orientation is not already in the orientations list and we push it into the list
* or it is already in the list and we remove it using this "filter" call that make sure that all instances (in case several were - by mistake? - available) of this orientation are removed (we keep only the values that are not equal to the orientation we want to remoev).

tsauerwein replied:
So `setOrientation` removes or adds an orientation? The name might be confusing. What about `toggleOrientation`?

asaunier replied:
Right! Done at dfc2490

tsauerwein commented on c2corg_ui/static/js/utils.js:
```diff
@@ -26,14 +26,17 @@ app.utils.getDoctype = function(type) {
-    .replace('{document_type}', document_type)
+app.utils.buildDocumentUrl = function(documentType, documentId, lang) {
+  if (!documentType || !documentId || !lang) {
+    return '';
```
I would rather return `null` in this case.

asaunier replied:
Doing so implies changing the "return" type of the function + the type of all vars that are set with this function. That probably makes sense but this PR already does that many changes, I would rather avoid adding one more. To do in another PR?

tsauerwein replied:
> To do in another PR?

Sure. Returning `null` also has the advantage that you are correctly handling the case that no url is returned wherever this function is called.

asaunier replied:
I agree, but I would prefer not having to also do that kind of changes in this PR.

tsauerwein commented:
I left some minor comments, feel free to merge if you want to do the rest in follow-up PRs.

asaunier commented:
> I left some minor comments, feel free to merge if you want to do the rest in follow-up PRs.

Thanks for reviewing @tsauerwein !

I have addressed most of your suggestions. I have also added most of the filters for routes. I think I will merge when Travis passes and continue with another PR for the rest of the route filters + the outing filters.
